python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll

```diff
@@ -0,0 +1,172 @@
+  private API::Node sendgridMailInstance() { result = sendgridMailHelper().getMember("Mail") }
+
+  /** Gets a call to `sendgrid.helpers.mail.Mail()`. */
+  private DataFlow::CallCfgNode sendgridMailCall() { result = sendgridMailInstance().getACall() }
+
+  /** Gets a reference to a `SendGridAPIClient` instance. */
+  private DataFlow::LocalSourceNode sendgridApiClient(DataFlow::TypeTracker t) {
+    t.start() and
+    result.(DataFlow::AttrRead).getObject*().getALocalSource() =
+      sendgrid().getMember("SendGridAPIClient").getReturn().getAUse()
+    or
+    exists(DataFlow::TypeTracker t2 | result = sendgridApiClient(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `SendGridAPIClient` instance use. */
+  private DataFlow::Node sendgridApiClient() {
+    sendgridApiClient(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /** Gets a reference to a `SendGridAPIClient` instance call with `send` or `post`. */
+  private DataFlow::Node sendgridApiSendCall() {
+    result = sendgridApiClient() and
+    result.(DataFlow::AttrRead).getAttributeName() in ["send", "post"]
+  }
+
+  private DataFlow::Node sendgridContent(DataFlow::CallCfgNode contentCall, string mime) {
+    exists(StrConst mimeNode |
+      mimeNode.getText() = mime and
+      DataFlow::exprNode(mimeNode).(DataFlow::LocalSourceNode).flowsTo(contentCall.getArg(0)) and
+      result = contentCall.getArg(1)
+    )
+  }
+
+  private DataFlow::Node sendgridWrite(string attributeName) {
+    exists(DataFlow::AttrWrite attrWrite |
+      attrWrite.getObject().getALocalSource() = sendgridMailCall() and
+      attrWrite.getAttributeName() = attributeName and
+      result = attrWrite.getValue()
+    )
+  }
+
+  /**
+   * Gets a reference to `sg.send()` and `sg.client.mail.send.post()`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   *  from_email = Email("from@example.com")
+   *  to_email = To("to@example.com")
+   *  subject = "Sending with SendGrid is Fun"
+   *  content = Content("text/html", request.args["html_content"])
+   *
+   *  mail = Mail(from_email, to_email, subject, content)
+   *
+   *  sg = SendGridAPIClient(api_key='SENDGRID_API_KEY')
+   *  response = sg.client.mail.send.post(request_body=mail.get())
+   *  ```
+   *
+   * * `this` would be `sg.client.mail.send.post(request_body=mail.get())`.
+   * * `getPlainTextBody()`'s result would be `none()`.
+   * * `getHtmlBody()`'s result would be `request.args["html_content"]`.
+   * * `getTo()`'s result would be `"to@example.com"`.
+   * * `getFrom()`'s result would be `"from@example.com"`.
+   * * `getSubject()`'s result would be `"Sending with SendGrid is Fun"`.
+   */
+  private class SendGridMail extends DataFlow::CallCfgNode, EmailSender::Range {
+    SendGridMail() { this.getFunction() = sendgridApiSendCall() }
+
+    override DataFlow::Node getPlainTextBody() {
+      result in [
+          sendgridMailCall().getArg(3), sendgridMailCall().getArgByName("plain_text_content")
```

Comment:
This will not synchronise properly. If you do not refer to `this`, you are reporting all plaintext bodies of all SendGrid calls here. It does not really affect the query as it just defines a set of sinks as being all HTML-bodies of all calls. But is does affect the modelling and I think it will also blow up the memory usage.